clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -3889,24 +4116,196 @@ scheduler_tmpl_t<RecordType, ReaderType>::set_output_active(output_ordinal_t out
+                unscheduled_priority_.queue.pop();
+            }
+        }
+        for (input_ordinal_t input : inputs_to_add) {
+            std::lock_guard<mutex_dbg_owned> lock(*inputs_[input].lock);
+            inputs_[input].unscheduled = false;
+        }
+    }
+
+    int live_inputs = live_input_count_.load(std::memory_order_acquire);
+    int live_outputs = 0;
+    for (unsigned int i = 0; i < outputs_.size(); ++i) {
+        if (outputs_[i].active->load(std::memory_order_acquire))
+            ++live_outputs;
+    }
+    double avg_per_output = live_inputs / static_cast<double>(live_outputs);
+    unsigned int avg_ceiling = static_cast<int>(std::ceil(avg_per_output));
+    unsigned int avg_floor = static_cast<int>(std::floor(avg_per_output));
+    int iteration = 0;
+    do {
+        // Walk the outputs, filling too-short queues from inputs_to_add and
+        // shrinking too-long queues into inputs_to_add.  We may need a 2nd pass
+        // for this; and a 3rd pass if bindings prevent even splitting.
+        VPRINT(
+            this, 3,
+            "Rebalance iteration %d inputs_to_add size=%zu avg_per_output=%4.1f %d-%d\n",
+            iteration, inputs_to_add.size(), avg_per_output, avg_floor, avg_ceiling);
+        // We're giving up the output locks as we go, so there may be some stealing
+        // in the middle of our operation, but the rebalancing is approximate anyway.
+        for (unsigned int i = 0; i < outputs_.size(); ++i) {
+            if (!outputs_[i].active->load(std::memory_order_acquire))
+                continue;
+            auto lock = acquire_scoped_output_lock_if_necessary(i);
+            // Only remove on the 1st iteration; later we can exceed due to binding
+            // constraints.
+            while (iteration == 0 && outputs_[i].runqueue.queue.size() > avg_ceiling) {
+                input_info_t *queue_next = nullptr;
+                // We use our regular pop_from_ready_queue which means we leave
+                // blocked inputs on the queue: those do not get rebalanced.
+                // XXX: Should we revisit that?
+                //
+                // We remove from the back to avoid penalizing the next-to-run entries
+                // at the front of the queue by putting them at the back of another
+                // queue.
+                status = pop_from_ready_queue_hold_locks(i, INVALID_OUTPUT_ORDINAL,
+                                                         queue_next, /*from_back=*/true);
+                if (status == STATUS_OK && queue_next != nullptr) {
+                    VPRINT(this, 3,
+                           "Rebalance iteration %d: output %d giving up input %d\n",
+                           iteration, i, queue_next->index);
+                    assert(status == STATUS_OK);
+                    inputs_to_add.push_back(queue_next->index);
+                } else {
+                    assert(status != STATUS_OK);
```

Comment:
I don't quite get this assert or the one above it (line 4265).  The first one seems superfluous, while this one appears to be asserting that STATUS_OK *implies* queue_next != nullptr.  But if that's the point, wouldn't it make more sense to move the queue_next check from the if () condition into an assert() that replaces the current assert(status==STATUS_OK)?

Reply:
Like the earlier redundant assert I believe my original code only checked for the queue_next not being null and asserted but it has to check the stats too.  Removed the redundant assert.

The other assert: STATUS_IDLE is often returned when queue_next==nullptr but not always so it seems incorrect actually; though I must never have hit that case -- which makes sense since this loop only tries to remove when there's a lot in the queue and that case only happens when the queue is empty.  In any case I removed the 2nd assert as well as it's adding more confusion than it's helping.